Jetpack Section: Backup Screen - Introduce Feature Flag

ParaskP7 opened the pull request:
Parent #13629 

This PR introduces a local feature flag for the `Backups` feature. When the feature flag is enabled a `Backup` menu is shown on the `My Site` tab (currently irrespective of the site plan).

To test:

- Launch app and go to `My Site` tab.
- Notice that `Backup` menu is hidden.
- Go to `My Site` -> `Toolbar Avatar` -> `App Settings` -> `Test feature configuration`.
- Enable `BackupsFeatureConfig`, scroll down and click the `RESTART THE APP` button. **DO NOT** confuse this feature flag with the `BackupFeatureConfig` (singular). Make sure to enable the right feature flag.
- Go back to `My Site` tab and notice `Backup` menu is shown.

<img src="https://user-images.githubusercontent.com/9729923/103663387-b5859600-4f79-11eb-85f0-8110a5a3c1ff.png" width="250" height="500">

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ParaskP7 committed:
Feature: Introduce backups feature flag

ParaskP7 committed:
Feature: Add backup menu on my site fragment

The backup menu is being displayed based on the backups feature config.

PS: The backup drawable is temporary and will be replaced with a
permanent one once the designs are finalized.

ParaskP7 committed:
Refactor: Optimize android R related imports for my site fragment

This fixes lots of import related warning on this class, thus making it
easier to spot a new warning.

ashiagr commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -1037,6 +1037,7 @@
     <string name="activity_log_item_menu_download_backup_label">Download backup</string>
 
     <!-- scan -->
+    <string name="backup">Backup</string>
```
Minor 🔍: We might want to put `Backup` related strings in a separate section.

ParaskP7 replied:
This is now done: 87590d6338a2c676f8860eafe5e818a595e6f9af

Thanks for point it out Ashita! ❤️

ashiagr reviewed (commented):
LGTM @ParaskP7! Also thank you for fixing android `R` imports 🙏.

Just a note for the future implementation: 

(I understand this PR only targets the "old My Site" section.)

We have a `ImprovedMySiteFragment` which will replace the old `MySiteFragment` soon. It gets enabled when `MySiteImprovementsFeatureConfig` is enabled in App Settings. 

"Scan" menu is functional only from the improved my site flow right now. We might want to sync in the next step to decide if we want to support new menus on the old My Site as I am not sure when `MySiteImprovements` is getting released.

ParaskP7 committed:
Refactor: Move newly added backup string to a separate section

ParaskP7 commented:
👋 @ashiagr !

Thank you so much for this review! 🌟 

> We have a ImprovedMySiteFragment which will replace the old MySiteFragment soon. It gets enabled when MySiteImprovementsFeatureConfig is enabled in App Settings.

You are totally right, I wasn't aware of that and missed that completely. I am currently updating this PR to add the `Backup` menu there as well. ⚙️ 

> "Scan" menu is functional only from the improved my site flow right now. We might want to sync in the next step to decide if we want to support new menus on the old My Site as I am not sure when MySiteImprovements is getting released.

- I am seeing `Scan` being functional from the `MySiteFragment` as well, what do you mean by it being functional only from the `ImprovedMySiteFragment`, am I missing something?
- I agree, we should definitely sync on whether we need to support new menus on the old `MySiteFragment` and when the new `ImprovedMySiteFragment` is going live.

ashiagr commented:
> I am seeing Scan being functional from the MySiteFragment as well, what do you mean by it being functional only from the ImprovedMySiteFragment, am I missing something?

Oh I thought Scan menu  is not connected to the Scan screen if accessed from the `MySiteFragment`; it is very-much connected, so we're good here. It's the display logic based on scan availability which is present only in the `ImprovedMySiteFragment's` VM. This logic is being replaced by the site capabilities API that we recently figured out ( @malinajirka is working on it). Let's sync on Slack so we're all on the same page.

ParaskP7 committed:
Feature: Add build backup site list item to site list item builder

The build backup list item functionality is based on the backups
feature config.

PS: The backup drawable is temporary and will be replaced with a
permanent one once the designs are finalized.

ParaskP7 committed:
Feature: Add build backup site list item to site items builder

The build backup list item functionality of the 'SiteItemsBuilder' is
triggered based on the backups feature config.

ParaskP7 committed:
Refactor: Move site items builder build func params to separate lines

This is because of the fact that the to be added new parameter will not
otherwise fit the line.

ParaskP7 committed:
Feature: Show backup site list item based on backups feature flag

ParaskP7 committed:
Feature: Bind backup site list menu item with site navigation

ParaskP7 committed:
Feature: Add a click listener to backup row for site navigation

This is an equivalent to what the previous commit did for the new
'ImprovedMySiteFragment'. As such, this commit is added to keep both
screens, the old and the new one, consistent with each other.

ParaskP7 committed:
Test: Refactor test to separate the arrange from the act

ParaskP7 committed:
Test: Optimize android R related imports for my site view model test

ParaskP7 committed:
Test: Use property access syntax on my site view model test

ParaskP7 committed:
Refactor: Reformat site items add all line on my site view model

ParaskP7 committed:
Test: Add missing any object matcher for my site view model tests

ParaskP7 commented:
👋 @ashiagr !

I have uploaded the solution for `ImprovedMySiteFragment` as well. I just re-requested your review.

Many thanks in advance! 🙏

ashiagr reviewed (approved):
Thanks for the changes @ParaskP7! I have verified that backup menu is present on both old and new My Site fragment based on the feature flag & currently irrespective of the site capabilities.

Also thanks for splitting code changes into separate commits and including appropriate information in the commit messages. It made reviewing very easy, good job 🌟!

ParaskP7 commented:
Thanks for approving and merging the PR @ashiagr ! 🌟 

I am really glad that you liked the way I structured my commits. I have been using this approach for few years now and been advocating in favor of it. It is really nice to see you found it valuable. In my opinion a reviewer should be aware of every step of the solution, small descriptive commits along with doing commit-by-commit review help navigating to the final outcome and in a sense doing an async pair programming with the author. I feel that this works really good. Also, doing that it is very beneficial for the author itself since they don't get overwhelmed by a major commit, which might include lots of feature additions, along with refactors, maybe tests, etc. With small commits, the author can quickly review their own commit much easier before committing it.

PS: Apologies for my mumbling above 😄 I am just really glad you liked it, and thanks again! 🤗
